crates/turborepo-ui/src/tui/app.rs

```diff
@@ -7,102 +8,265 @@ use std::{
-        let Some(selected_task) = self.table.selected() else {
-            return;
-        };
-        if self.pane.has_stdin(selected_task) {
-            self.input_options.interact = interact;
-            self.pane.highlight(interact);
+    /// Mark the given running task as finished
+    /// Errors if given task wasn't a running task
+    pub fn finish_task(&mut self, task: &str, result: TaskResult) -> Result<(), Error> {
+        // Name of currently highlighted task.
+        // We will use this after the order switches.
+        let highlighted_task =
+            &self.tasks_by_status.task_names_in_displayed_order()[self.selected_task_index];
+
+        let running_idx = self
+            .tasks_by_status
+            .running
+            .iter()
+            .position(|running| running.name() == task)
+            .ok_or_else(|| {
+                debug!("could not find '{task}' to finish");
+                println!("{:#?}", highlighted_task);
+                Error::TaskNotFound { name: task.into() }
+            })?;
+
+        let running = self.tasks_by_status.running.remove(running_idx);
+        self.tasks_by_status.finished.push(running.finish(result));
+
+        // If user hasn't interacted, keep highlighting top-most task in list.
+        if !self.has_user_scrolled {
+            return Ok(());
         }
+
+        // Find the highlighted task from before the list movement in the new list.
+        if let Some(new_index_to_highlight) = self
+            .tasks_by_status
+            .task_names_in_displayed_order()
+            .iter()
+            .position(|running| running == highlighted_task)
+        {
+            self.selected_task_index = new_index_to_highlight;
+            self.scroll.select(Some(new_index_to_highlight));
+        }
+
+        Ok(())
     }
 
-    pub fn scroll(&mut self, direction: Direction) {
-        let Some(selected_task) = self.table.selected() else {
-            return;
-        };
-        self.pane
-            .scroll(selected_task, direction)
-            .expect("selected task should be in pane");
+    pub fn has_stdin(&self) -> bool {
+        let active_task = self.active_task();
+        if let Some(term) = self.tasks.get(&active_task) {
+            term.stdin.is_some()
+        } else {
+            false
+        }
     }
 
-    pub fn term_size(&self) -> (u16, u16) {
-        self.pane.term_size()
+    pub fn interact(&mut self) {
+        if matches!(self.input_options.focus, LayoutSections::Pane) {
+            self.input_options.focus = LayoutSections::TaskList
+        } else if self.has_stdin() {
+            self.input_options.focus = LayoutSections::Pane;
+        }
     }
 
     pub fn update_tasks(&mut self, tasks: Vec<String>) {
-        self.table = TaskTable::new(tasks.clone());
-        self.next();
+        let mut task_list = tasks.into_iter().map(Task::new).collect::<Vec<_>>();
+        task_list.sort_unstable();
+        task_list.dedup();
```

Comment:
This function needs to get finished. We're not saving the new list anywhere. We should be:
 - Overwriting `task_by_status` for the list of tasks
 - Make sure all new task names have an entry in `tasks` so we don't crash when they produce output